cpp/dolfin/mesh/MeshValueCollection.h

```diff
@@ -184,6 +189,98 @@ MeshValueCollection<T>::MeshValueCollection(std::shared_ptr<const Mesh> mesh,
 {
   // Do nothing
 }
+
+
+template <typename T>
+MeshValueCollection<T>::MeshValueCollection(
+                      std::shared_ptr<const Mesh> mesh,
+                      std::size_t dim,
+                      std::vector<std::vector<T>>& cells,
+                      std::vector<T>& values_data): _mesh(mesh), _dim(dim)
+{ 
+  // Ensure the mesh dimension is initialised. 
+  // If the mesh is created from arrays, entities 
+  // of all dimesions are not initialized.
+  _mesh->create_entities(dim);
+
+  // The number of vertices per entity is _dim+1
+  std::vector<std::int32_t> v(_dim+1);
+
+  // Map from {entity vertex indices} to entity index
+  std::map<std::vector<int>, size_t> entity_map;
+
+  const std::vector<std::int64_t>& global_indices
+      = _mesh->topology().global_indices(0);
+
+  // Loop over all the entities of dimension _dim
+  for (auto& m : mesh::MeshRange(*mesh, _dim))
+  { 
+    if (_dim == 0)
+      v[0] = global_indices[m.index()];
+    else
+    {
+      v.clear();
+      for (auto& vtx : mesh::EntityRange(m, 0)){
+        v.push_back(global_indices[vtx.index()]);
+      }
+      std::sort(v.begin(), v.end());
+    }
+    // The vector of vertex number is key and entity index is value
+    entity_map[v]=m.index();
+  }
+
+  const std::size_t D = _mesh->topology().dim();
+
+  // Handle cells as a special case
+  if ((int)D == _dim)
+  {
+    for (std::size_t cell_index = 0; cell_index < values_data.size();
+         ++cell_index)
+    {
+      const std::pair<std::size_t, std::size_t> key(cell_index, 0);
+      _values.insert({key, values_data[cell_index]});
+    }
+  }
+  else
+  {
+    _mesh->create_connectivity(_dim, D);
+
+    assert(_mesh->topology().connectivity(_dim, D));
+    const Connectivity& connectivity = *_mesh->topology().connectivity(_dim, D);
+
+    for (std::size_t j = 0; j < values_data.size();
+         ++j)
+    {
+      // Find the cell
+      v.clear();
+      // cells[j] is a vector of length _dim+1
+      for (std::size_t i = 0; i < _dim+1; ++i){
```

Comment:
Wrong assumption on number of vertices per cell here again.

Reply:
updated

Reply:
corrected